sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs

```diff
@@ -1054,5 +1056,291 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
+            {
+                scope.Failed(ex);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Retrieve raw events for a given Time Series Id over a certain timeSpan synchronously.
+        /// </summary>
+        /// <param name="timeSeriesId">The Time Series Id to retrieve raw events for.</param>
+        /// <param name="timeSpan">The timespan over which to query data.</param>
+        /// <param name="endTime">End timestamp of the time range. Events that match this timestamp are excluded. If null is provided, DateTimeOffset.UtcNow is used.</param>
+        /// <param name="options">Optional parameters to use when querying for events.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns>The pageable list <see cref="AsyncPageable{QueryResultPage}"/> of query result frames.</returns>
+        public virtual Pageable<QueryResultPage> QueryEvents(
+            TimeSeriesId timeSeriesId,
+            TimeSpan timeSpan,
+            DateTimeOffset? endTime = null,
+            QueryEventsRequestOptions options = null,
+            CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(TimeSeriesInsightsClient)}.{nameof(QueryEvents)}");
+            scope.Start();
+
+            try
+            {
+                DateTimeOffset rangeEndTime = endTime ?? DateTimeOffset.UtcNow;
+                DateTimeOffset rangeStartTime = rangeEndTime - timeSpan;
+                return QueryEventsInternal(timeSeriesId, rangeStartTime, rangeEndTime, options, cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+        }
+
+        private AsyncPageable<QueryResultPage> QueryEventsInternalAsync(
+            TimeSeriesId timeSeriesId,
+            DateTimeOffset startTime,
+            DateTimeOffset endTime,
+            QueryEventsRequestOptions options,
+            CancellationToken cancellationToken)
+        {
+            var searchSpan = new DateTimeRange(startTime, endTime);
+            var queryRequest = new QueryRequest
+            {
+                GetEvents = new GetEvents(timeSeriesId, searchSpan)
+            };
+
+            if (options != null)
+            {
+                if (options.Filter != null)
+                {
+                    queryRequest.GetEvents.Filter = new TimeSeriesExpression(options.Filter);
+                }
+
+                if (options.ProjectedProperties != null)
+                {
+                    foreach (EventProperty projectedProperty in options.ProjectedProperties)
+                    {
+                        queryRequest.GetEvents.ProjectedProperties.Add(projectedProperty);
+                    }
+                }
+
+                queryRequest.GetEvents.Take = options.MaximumNumberOfEvents;
+            }
```

Comment:
This is repeated below, it might make sense to make a private function out of it.